Frameworks/UIKit/XamlControls.h

```diff
@@ -26,118 +26,97 @@
+#import <winrt/Windows.UI.Xaml.Input.h>
 #include <COMIncludes_end.h>
 
-#import <UWP/WindowsUIXamlControls.h>
-#import <UWP/WindowsUIXamlInput.h>
-
 namespace XamlControls {
 
 ////////////////////////////////////////////////////////////////////////////////////
 // Shared Helpers
 ////////////////////////////////////////////////////////////////////////////////////
 #if !__has_feature(objc_arc)
-// TODO: GitHub issue 508 and 509
-// We need a type-safe way to do this with projections.  This is copied verbatim from the projections
-// code and works perfectly for this limited usage, but we don't do any type validation below.
-inline id _createRtProxy(Class cls, IInspectable* iface) {
-    // Oddly, WinRT can hand us back NULL objects from successful function calls. Plumb these through as nil.
-    if (!iface) {
-        return nil;
-    }
-
-    RTObject* ret = [NSAllocateObject(cls, 0, 0) init];
-    [ret setComObj:iface];
-
-    return [ret autorelease];
-}
-
-inline RTObject* _createBareRTObj(IInspectable* obj) {
-    return _createRtProxy(objc_getClass("RTObject"), obj);
-}
-
-template <typename I>
-inline I* _createRtProxy(IInspectable* iface) {
-    return _createRtProxy([I class], iface);
-}
-
 // TODO: Remove when we move our controls over to actual XAML-build control implementations
 class WUXIPointerEventHandler_shim
     : public Microsoft::WRL::RuntimeClass<Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::WinRtClassicComMix>,
                                           ABI::Windows::UI::Xaml::Input::IPointerEventHandler> {
-    void (^_delegate)(RTObject*, WUXIPointerRoutedEventArgs*);
+    winrt::Windows::UI::Xaml::Input::PointerEventHandler _handler;
 
 public:
-    WUXIPointerEventHandler_shim(void (^del)(RTObject*, WUXIPointerRoutedEventArgs*)) : _delegate([del copy]) {
-    }
-    ~WUXIPointerEventHandler_shim() {
-        [_delegate release];
+    WUXIPointerEventHandler_shim(const winrt::Windows::UI::Xaml::Input::PointerEventHandler& handler) : _handler(handler) {
     }
     virtual HRESULT STDMETHODCALLTYPE GetTrustLevel(TrustLevel* trustLevel) override {
         *trustLevel = BaseTrust;
         return S_OK;
     }
     virtual HRESULT STDMETHODCALLTYPE Invoke(IInspectable* arg0, ABI::Windows::UI::Xaml::Input::IPointerRoutedEventArgs* arg1) override {
-        NSAutoreleasePool* p = [NSAutoreleasePool new];
-        _delegate(_createBareRTObj(arg0), _createRtProxy<WUXIPointerRoutedEventArgs>(arg1));
+        winrt::Windows::Foundation::IInspectable sender = nullptr;
+        winrt::copy_from_abi(sender, reinterpret_cast<winrt::ABI::Windows::Foundation::IInspectable*>(arg0));
 
+        winrt::Windows::UI::Xaml::Input::PointerRoutedEventArgs args = nullptr;
+        winrt::copy_from_abi(args, reinterpret_cast<winrt::ABI::Windows::UI::Xaml::Input::IPointerRoutedEventArgs*>(arg1));
+
+        NSAutoreleasePool* p = [NSAutoreleasePool new];
+        _handler(sender, args);
         [p release];
```

Comment:
This may also be a good time to move to @autoreleasepool or lower-level push/pop calls vs. allocating an NSAutoReleasePool every time.

---
In reply to: [109247283](https://github.com/Microsoft/WinObjC/pull/2382#discussion_r109247283) [](ancestors = 109247283)